File: scripts/test_solutions/269_alien_dictionary.js
var alienOrder = function(words) {
    const graph = new Map(), indegree = new Map();
    for (const w of words) for (const c of w) { graph.set(c, new Set()); indegree.set(c, 0); }
    for (let i = 0; i < words.length - 1; i++) {
        const w1 = words[i], w2 = words[i + 1];
        if (w1.length > w2.length && w1.startsWith(w2)) return '';
        for (let j = 0; j < Math.min(w1.length, w2.length); j++) {
            if (w1[j] !== w2[j]) {
                if (!graph.get(w1[j]).has(w2[j])) { graph.get(w1[j]).add(w2[j]); indegree.set(w2[j], indegree.get(w2[j]) + 1); }
                break;
            }
        }
    }
    const queue = [...indegree.keys()].filter(c => indegree.get(c) === 0), result = [];
    while (queue.length) {
        const c = queue.shift(); result.push(c);
        for (const next of graph.get(c)) { indegree.set(next, indegree.get(next) - 1); if (indegree.get(next) === 0) queue.push(next); }
    }
    return result.length === graph.size ? result.join('') : '';
};

File: scripts/test_solutions/269_alien_dictionary.py
from collections import defaultdict, deque

def alienOrder(words):
    graph = defaultdict(set)
    indegree = {c: 0 for w in words for c in w}

    for i in range(len(words) - 1):
        w1, w2 = words[i], words[i + 1]
        if len(w1) > len(w2) and w1.startswith(w2):
            return ''
        for j in range(min(len(w1), len(w2))):
            if w1[j] != w2[j]:
                if w2[j] not in graph[w1[j]]:
                    graph[w1[j]].add(w2[j])
                    indegree[w2[j]] += 1
                break

    queue = deque([c for c in indegree if indegree[c] == 0])
    result = []
    while queue:
        c = queue.popleft()
        result.append(c)
        for next_c in graph[c]:
            indegree[next_c] -= 1
            if indegree[next_c] == 0:
                queue.append(next_c)

    return ''.join(result) if len(result) == len(indegree) else ''
